bpf/process/bpf_rate.h

```diff
@@ -0,0 +1,61 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+/* Copyright Authors of Cilium */
+
+#ifndef __RATE_H__
+#define __RATE_H__
+
+#include "bpf_tracing.h"
+
+struct cgroup_rate_key {
+	__u8 op;
+	__u8 pad[7];
+	__u64 cgroupid;
+};
+
+struct cgroup_rate_value {
+	__u64 lastns;
+	__u64 tokens;
+};
+
+struct cgroup_rate_settings {
+	__u64 tokens;
+	__u64 interval_ns;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 32768);
+	__type(key, struct cgroup_rate_key);
+	__type(value, struct cgroup_rate_value);
+} cgroup_rate_map SEC(".maps");
+
+static inline __attribute__((always_inline)) bool
+cgroup_rate(struct cgroup_rate_key *key, __u64 ns,
+	    const struct cgroup_rate_settings *settings)
+{
+	struct cgroup_rate_value *value, new_value;
+	__u64 since_lastns;
+
+	value = map_lookup_elem(&cgroup_rate_map, key);
+	if (!value) {
+		new_value.lastns = ns;
+		new_value.tokens = settings->tokens - 1;
+		map_update_elem(&cgroup_rate_map, key, &new_value, BPF_ANY);
+		return true;
+	}
+
+	since_lastns = ns - value->lastns;
+	if (since_lastns > settings->interval_ns) {
+		value->tokens = settings->tokens;
+		value->lastns = ns;
```

Comment:
hm right, but I think it needs to be global - per cgroup ... at least I'm not sure how the per-core would work

I have another way of doing this but that involves timers (since 5.15) but should be sync safe ... I'll try that one

Reply:
The idea would be that every core would have a map for each cgroup. Maybe I'm missing something, but it should be exactly the same code with the only difference being that the hashmap is pser-cpu.

Reply:
but we need per cgroup rate.. this would spread it across cpus, I might be missing something ;-)

Reply:
Hmm... Maybe I'm missing something.

The current implementation seems to use a global rate across all CPUs of the system. So if rate is X, the global event rate of a given cgroup cannot exceed X.

I was thinking of the rates being per-core rather than global (system-wide). So if rate is X, this would mean that the event rate of a given cgroup cannot exceed X on any given CPU. Note that this bounds the system rate into X*<ncpus> system-wide.

On problem with the per-core idea, however, is the cleanup. Because when a cgroup is removed in one core, you need to remove the entries from all of them, which (I think) means that you would need to do the removal in user-space.